## PешАлово 2.0 или Что нужно государству?


Author:  olga.zagovora@gesis.org
    
Data from Serhiy Protsenko here https://github.com/SerhiyProtsenko/Magister_2018_K1-K9 

In [2]:
from collections import defaultdict
import pandas as pd
import os
import fnmatch
import json
import sys
stdout = sys.stdout
reload(sys)
sys.setdefaultencoding('utf-8')
sys.stdout = stdout

In [165]:
df=pd.read_excel("Mag2018_K1_k9.xlsx")
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3014 entries, 1307 to 38
Data columns (total 23 columns):
рік                                                         3014 non-null int64
код ЗВО                                                     3014 non-null int64
Назва ЗВО                                                   3014 non-null object
Код спец.                                                   3014 non-null int64
Код. спеціал                                                3014 non-null int64
Назва спец.                                                 3014 non-null object
K1                                                          3014 non-null float64
K2                                                          3014 non-null float64
K3                                                          3014 non-null float64
K4                                                          3014 non-null float64
K5                                                          3014 non-null float64


In [166]:
import statsmodels.formula.api as smf
import statsmodels.api as sm

In [167]:
just_dummies = pd.get_dummies(df[u'Назва ЗВО'])#,drop_first=True)
df_1 = pd.concat([df, just_dummies], axis=1)      
#df_1.drop(['Назва ЗВО'], inplace=True, axis=1)
df_1.rename(columns={u'Обсяг державного замовлення Всього, денна':"ODZ",u'Mult K1-K9':"Mult_K1_K9",
                     u'Випуск бак., денна':"VB",u'Випуск бак. min 5, денна':"VB5",
                     u'Назва ЗВО':"Nazva_ZVO",u'Назва спец.':"Nazva_spec"}, inplace=True)
df_2=df_1[[u'Mult_K1_K9',"VB","ODZ"]]
df_2=sm.add_constant(df_2)

In [160]:
y=df_1["ODZ"]

In [161]:
result = smf.ols(formula='ODZ ~ Mult_K1_K9 + VB', data=df_1).fit()
#result = sm.OLS(y,df_1[["const","Mult_K1_K9" ,"VB"]]).fit()
print result.summary()

                            OLS Regression Results                            
Dep. Variable:                    ODZ   R-squared:                       0.845
Model:                            OLS   Adj. R-squared:                  0.845
Method:                 Least Squares   F-statistic:                     8194.
Date:                Mon, 23 Jul 2018   Prob (F-statistic):               0.00
Time:                        14:55:54   Log-Likelihood:                -10446.
No. Observations:                3014   AIC:                         2.090e+04
Df Residuals:                    3011   BIC:                         2.092e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -8.5688      1.076     -7.967      0.0

In [112]:
df_1[["Nazva_ZVO",u'Mult_K1_K9',u'ODZ',"VB"]].head()

Nazva_ZVO  Mult_K1_K9  ODZ  VB
1307  Донецький нацiональний унiверситет ім. Василя ...    2.212647    6   3
2061  Київський національний університет ім. Тараса ...    2.075149   36  56
2419  Харківський національний унiверситет iм. В.Н.К...    1.960006   16  16
634                      Сумський державний унiверситет    1.872965   11  11
1569  Львівський нацiональний унiверситет імені Іван...    1.855975   21  23

In [94]:
7.55*1.87+0.5*11-8.56 #R^2 0.845

11.0585

In [140]:
result = smf.ols(formula='ODZ ~ Mult_K1_K9 + VB5', data=df_1).fit()
print result.summary()

                            OLS Regression Results                            
Dep. Variable:                    ODZ   R-squared:                       0.848
Model:                            OLS   Adj. R-squared:                  0.848
Method:                 Least Squares   F-statistic:                     8415.
Date:                Mon, 23 Jul 2018   Prob (F-statistic):               0.00
Time:                        14:41:21   Log-Likelihood:                -10412.
No. Observations:                3014   AIC:                         2.083e+04
Df Residuals:                    3011   BIC:                         2.085e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -9.4029      1.063     -8.846      0.0

In [96]:
result = smf.ols(formula='ODZ ~ Mult_K1_K9*VB', data=df_1).fit()
print result.summary()

                            OLS Regression Results                            
Dep. Variable:                    ODZ   R-squared:                       0.861
Model:                            OLS   Adj. R-squared:                  0.861
Method:                 Least Squares   F-statistic:                     6210.
Date:                Mon, 23 Jul 2018   Prob (F-statistic):               0.00
Time:                        14:14:39   Log-Likelihood:                -10281.
No. Observations:                3014   AIC:                         2.057e+04
Df Residuals:                    3010   BIC:                         2.059e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept         6.2630      1.291      4.850

In [134]:
0.3979*1.87*11 - 0.1235*11 + 6.26    #R^2 0.861

13.086303000000001

In [141]:
result = smf.ols(formula='ODZ ~ Mult_K1_K9*VB5', data=df_1).fit()
print result.summary()

                            OLS Regression Results                            
Dep. Variable:                    ODZ   R-squared:                       0.863
Model:                            OLS   Adj. R-squared:                  0.863
Method:                 Least Squares   F-statistic:                     6312.
Date:                Mon, 23 Jul 2018   Prob (F-statistic):               0.00
Time:                        14:42:15   Log-Likelihood:                -10260.
No. Observations:                3014   AIC:                         2.053e+04
Df Residuals:                    3010   BIC:                         2.055e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept          5.4720      1.309      4.

In [184]:
df_1["Nazva_ZVO"]=df_1.Nazva_ZVO.apply(lambda x: x.split()[0]+" "+"".join(item[0].upper() for item in x.split()[1:]))

In [185]:
from transliterate import translit
df_1["Nazva_ZVO"]=df_1["Nazva_ZVO"].apply(lambda x: translit(x, "uk", reversed=True))

In [145]:
result = smf.ols(formula='ODZ ~ Mult_K1_K9*VB5 + C(Nazva_ZVO)', data=df_1).fit()
print result.summary()

                            OLS Regression Results                            
Dep. Variable:                    ODZ   R-squared:                       0.880
Model:                            OLS   Adj. R-squared:                  0.874
Method:                 Least Squares   F-statistic:                     146.9
Date:                Mon, 23 Jul 2018   Prob (F-statistic):               0.00
Time:                        14:43:07   Log-Likelihood:                -10061.
No. Observations:                3014   AIC:                         2.041e+04
Df Residuals:                    2870   BIC:                         2.128e+04
Df Model:                         143                                         
Covariance Type:            nonrobust                                         
                                                       coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------

In [155]:
#print 0.3430*1.87 * 11  + 6.2199
#print 0.3430*1.87 * 11 - 2.3647*1.87 -0.0284*11  + 6.2199

sumDU_k1k9_mev=df_1[df_1[u'Nazva_ZVO']=="Sums'kyj DU"][u'Mult_K1_K9'].iloc[0]
sumDU_bak_mev=df_1[df_1[u'Nazva_ZVO']=="Sums'kyj DU"][u'VB5'].iloc[0]

print 0.3430*sumDU_k1k9_mev * sumDU_bak_mev  + 6.2199
print 0.3430*sumDU_k1k9_mev *sumDU_bak_mev - 2.3647*sumDU_k1k9_mev -0.0284*sumDU_bak_mev  + 6.2199

13.2865962206
8.54519633911


In [156]:
df[df[u'Назва ЗВО']=="Сумський державний унiверситет"]#[u'Mult K1-K9']#.iloc[0]

рік  код ЗВО                       Назва ЗВО  Код спец.  Код. спеціал  \
634  2018      168  Сумський державний унiверситет        292         11111   
633  2018      168  Сумський державний унiверситет        281         11111   
632  2018      168  Сумський державний унiверситет        171         11111   
631  2018      168  Сумський державний унiверситет        163         11111   
630  2018      168  Сумський державний унiверситет        161         11111   
629  2018      168  Сумський державний унiверситет        153         11111   
628  2018      168  Сумський державний унiверситет        152         11111   
627  2018      168  Сумський державний унiверситет        151         11111   
626  2018      168  Сумський державний унiверситет        144         11111   
625  2018      168  Сумський державний унiверситет        142         11111   
624  2018      168  Сумський державний унiверситет        141         11111   
623  2018      168  Сумський державний унiверситет        133         11111   
622  2018      168  Сумський державний унiверситет        132         11111   
621  2018      168  Сумський державний унiверситет        131         11111   
620  2018      168  Сумський державний унiверситет        122         11111   
619  2018      168  Сумський державний унiверситет        113         11111   
618  2018      168  Сумський державний унiверситет        101         11111   
617  2018      168  Сумський державний унiверситет         76         11111   
616  2018      168  Сумський державний унiверситет         75         11111   
615  2018      168  Сумський державний унiверситет         73         11111   
614  2018      168  Сумський державний унiверситет         72         11111   
613  2018      168  Сумський державний унiверситет         71         11111   
612  2018      168  Сумський державний унiверситет         61         11111   
611  2018      168  Сумський державний унiверситет         51         11111   
610  2018      168  Сумський державний унiверситет         35            41   
609  2018      168  Сумський державний унiверситет         17         11111   

                                           Назва спец.    K1    K2   K3    K4  \
634                    Міжнародні економічні відносини  1.25  1.07  1.1  1.04   
633             Публічне управління та адміністрування  1.25  1.07  1.1  1.04   
632                                        Електроніка  1.25  1.07  1.1  1.04   
631                               Біомедична інженерія  1.25  1.07  1.1  1.04   
630                    Хімічні технології та інженерія  1.25  1.07  1.1  1.04   
629                     Мікро- та наносистемна техніка  1.25  1.07  1.1  1.04   
628    Метрологія та інформаційно-вимірювальна техніка  1.25  1.07  1.1  1.04   
627           Автоматизація та комп’ютерно-інтегровані  1.25  1.07  1.1  1.04   
626                                    Теплоенергетика  1.25  1.07  1.1  1.04   
625                        Енергетичне машинобудування  1.25  1.07  1.1  1.04   
624               Електроенергетика, електротехніка та  1.25  1.07  1.1  1.04   
623                           Галузеве машинобудування  1.25  1.07  1.1  1.04   
622                                  Матеріалознавство  1.25  1.07  1.1  1.04   
621                                 Прикладна механіка  1.25  1.07  1.1  1.04   
620                                  Комп’ютерні науки  1.25  1.07  1.1  1.04   
619                               Прикладна математика  1.25  1.07  1.1  1.04   
618                                           Екологія  1.25  1.07  1.1  1.04   
617     Підприємництво, торгівля та біржова діяльність  1.25  1.07  1.1  1.04   
616                                          Маркетинг  1.25  1.07  1.1  1.04   
615                                         Менеджмент  1.25  1.07  1.1  1.04   
614          Фінанси, банківська справа та страхування  1.25  1.07  1.1  1.04   
613                              Облік і оподаткування  1.25  1.07  1.1  1.04   
612      

In [181]:
df_1["Nazva_spec"]=df_1["Nazva_spec"].apply(lambda x: translit(x, "uk", reversed=True))

df_1["Nazva_spec"]=df_1.Nazva_spec.apply(lambda x: x.split()[0]+" "+"".join(item[:3].title() for item in x.split()[1:]))

In [186]:
result = smf.ols(formula='ODZ ~ Mult_K1_K9*VB5 + C(Nazva_ZVO)+ C(Nazva_spec)', data=df_1).fit()
print result.summary()

                            OLS Regression Results                            
Dep. Variable:                    ODZ   R-squared:                       0.904
Model:                            OLS   Adj. R-squared:                  0.893
Method:                 Least Squares   F-statistic:                     80.34
Date:                Mon, 23 Jul 2018   Prob (F-statistic):               0.00
Time:                        15:09:58   Log-Likelihood:                -9722.8
No. Observations:                3014   AIC:                         2.008e+04
Df Residuals:                    2697   BIC:                         2.199e+04
Df Model:                         316                                         
Covariance Type:            nonrobust                                         
                                                          coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------

In [189]:
print "SumDU, Міжнародні економічні відносини (C(Nazva_spec)[T.Mizhnarodni EkoVid])"
print 0.3508 *sumDU_k1k9_mev *sumDU_bak_mev -10.55 * 1  +  13.2888

SumDU, Міжнародні економічні відносини (C(Nazva_spec)[T.Mizhnarodni EkoVid])
9.96619660111


# Кому дали більше місць? І кого дискримінували?

In [213]:
s=result.pvalues
s_param01=result.params[result.params.index.isin(s[s<0.01].index)]

In [ ]:
# p val < 0.01

# ПорешАли?

In [208]:
s_param01[s_param01>0]

Intercept                                 13.288817
C(Nazva_ZVO)[T.L'otna AN]                 23.972899
C(Nazva_ZVO)[T.Natsional'nyj TUU"PIIS]     5.575062
C(Nazva_ZVO)[T.Tavrijs'kyj DAU]            7.555305
Mult_K1_K9:VB5                             0.350807
dtype: float64

### Льотна академiя НАУ

### Національний технiчний унiверситет України "Київський політехнічний інститут ім.Ігоря Сікорського"

### Таврійський державний агротехнологічний університет



In [ ]:
# p val < 0.05

In [211]:
s_param05=result.params[result.params.index.isin(s[s<0.05].index)]
s_param05[s_param05>0]

Intercept                                           13.288817
C(Nazva_ZVO)[T.Kryvoriz'kyj DPU]                     4.413021
C(Nazva_ZVO)[T.L'otna AN]                           23.972899
C(Nazva_ZVO)[T.L'vivs'ka NAM]                        8.328672
C(Nazva_ZVO)[T.Natsional'nyj TUU"PIIS]               5.575062
C(Nazva_ZVO)[T.Perejaslav-Khmel'nyts'kyj DPUIHS]     4.226128
C(Nazva_ZVO)[T.Tavrijs'kyj DAU]                      7.555305
C(Nazva_spec)[T.Aviatsijnyj Tra]                    10.213132
C(Nazva_spec)[T.Zaliznychnyj Tra]                    7.824164
Mult_K1_K9:VB5                                       0.350807
dtype: float64

### Криворiзький державний педагогiчний унiверситет

### Льотна академiя НАУ

### Львівська національна академiя мистецтв

### Національний технiчний унiверситет України "Київський політехнічний інститут ім.Ігоря Сікорського"

### Переяслав-Хмельницький державний педагогiчний унiверситет ім. Григорія Сковороди

### Таврійський державний агротехнологічний університет


### Спеціальності: Авіаційний транспорт, Залізничний транспорт

## Дискримінували (p<0.01)

In [209]:
s_param01[s_param01<0]

C(Nazva_ZVO)[T.Derzhavnyj UITT]                        -8.980964
C(Nazva_ZVO)[T.Kyyivs'kyj NTU]                        -14.336335
C(Nazva_ZVO)[T.Poltavs'kyj NPUIV]                      -4.850090
C(Nazva_spec)[T.Ahronomija ]                           -7.126173
C(Nazva_spec)[T.Avionika ]                            -10.671719
C(Nazva_spec)[T.Avtomatyzatsija TaKom]                -10.612061
C(Nazva_spec)[T.Avtomatyzatsija TaKomTek]              -9.648574
C(Nazva_spec)[T.Avtomobil'nyj Tra]                    -10.404352
C(Nazva_spec)[T.Biomedychna Inz]                      -12.387371
C(Nazva_spec)[T.Biotekhnolohiyi TaBio]                 -6.672218
C(Nazva_spec)[T.Doshkil'na Osv]                        -8.359933
C(Nazva_spec)[T.Dyzajn ]                              -15.107574
C(Nazva_spec)[T.Ekolohija ]                           -12.545108
C(Nazva_spec)[T.Ekonomika ]                           -14.638581
C(Nazva_spec)[T.Elektroenerhetyka, EleTa]             -10.171917
C(Nazva_spec)[T.Elektroen

Дискримінація вузів

C(Nazva_ZVO)[T.Derzhavnyj UITT]                        -8.980964

C(Nazva_ZVO)[T.Kyyivs'kyj NTU]                        -14.336335
             
C(Nazva_ZVO)[T.Poltavs'kyj NPUIV]                      -4.850090
             

### Державний університет інфраструктури та технологій
             
### Київський нацiональний торговельно-економiчний унiверситет
             
### Полтавський національний педагогiчний унiверситет ім.В.Г.Короленка
             
             
             
             